### PR TITLE
chrootenv: rewrite on top of GLib

### LINKED ISSUE #33106: chrootenv: breaks when root contains broken symlink
## Issue description
Steam no longer launches.
Reverting the changes introduced in #31182 and #32899 fixes the issue.

### Steps to reproduce
```
❯ $(nix-build '<nixpkgs>' -A steam)/bin/steam
/nix/store/iqh6if4haz9x510q0166mz0k55z26k8l-chrootenv:chrootenv.c:59: stat(/result): No such file or directory
/nix/store/iqh6if4haz9x510q0166mz0k55z26k8l-chrootenv:chrootenv.c:132: nftw(/tmp/chrootenvKFDgfE): Directory not empty
```

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.71, NixOS, 18.03.git.ade98dc (Impala)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.12pre5810_5d5b931f`
 - channels(root): `""`
 - channels(sarah): `""`
 - nixpkgs: `/home/sarah/src/nixpkgs`
```

### PR DESCRIPTION
**Pull request by lukateras**
Changes:

* doesn't handle root user separately
* doesn't `chdir("/")` which makes using it seamless
* only bind mounts, doesn't symlink

Incidentally, it fixes #33106.

It's about two times shorter than the previous version (it's even shorter than the Ruby version), and much, much easier to read/follow through. It uses GLib quite heavily, along with RAII (available in GCC/Clang).

Symlinking already happens in init script, so it's redundant to have it in this code. It has only been in chrootenv for a week, and I might reintroduce it back later.

About closure size: Steam itself uses GLib, so it won't result in a larger closure for most chrootenv users. It's also likely to be required by other programs. I've tested Steam, and of course, it still works.

/cc @Mic92 @dezgeg @srhb @pbogdan

### CODE REVIEW

**Comment by yegortimoshenko** [comment]
I feel confident about new code and will merge on 31st if no one has any issues with it.

I should note that none of the changes to chrootenv are breaking `buildFHSUserEnv` behavior because root-specific logic was an optimization, scripts that wrap chrootenv `chdir` and `symlink` anyway.

File: pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    if (nftw(prefix, nftw_remove, getdtablesize(),
+             FTW_DEPTH | FTW_MOUNT | FTW_PHYS) < 0)
+      g_perror("nftw");
+
+    if (WIFEXITED(status))
+      return WEXITSTATUS(status);
+
+    else if (WIFSIGNALED(status))
+      kill(getpid(), WTERMSIG(status));
+  }
```
Comment (orivej) [inline pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: It is not obvious that the equivalent of `return EX_OSERR;` of the old version is unneeded in the new version: the signal that terminated the child process may be ignored by default and not terminate chrootenv.

Comment (yegortimoshenko) [reply pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: If signal is ignored, it will return 0. This seems to be logical behavior.

Comment (orivej) [reply pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: If the child overrode the default of ignoring a signal and died on receiving it, it did not exit cleanly and chrootenv should not return 0.

Comment (yegortimoshenko) [reply pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: You are right, it's better to return 1 in this case. I've added it, see https://github.com/NixOS/nixpkgs/pull/33139/commits/ac40a57cea695493e0da2b91530688ea2da4f240.

**Comment by FRidh** [comment]
What do you think about documenting what this code is for and how it (roughly) works. I think each piece of code should explain why it is needed. So, what problem does this code solve?

Looking at the older version, there were some comments that helps the reader understand what is happening.

**Comment by yegortimoshenko** [comment]
There are only two comments that still apply, specifically "Create new mount and user namespaces" and "Map users and groups to the parent namespace". I am not sure if these are particularly useful to keep around, because people who know about mount and user namespaces probably know about `unshare` anyway. If not, then `man 2 unshare` provides better documentation than these comments.

> So, what problem does this code solve?

Setting up a chroot to ultimately run programs that assume a FHS-compliant environment.

**Comment by FRidh** [comment]
> Setting up a chroot to ultimately run programs that assume a FHS-compliant environment.

That at the very least should be included in the code header.

**Comment by yegortimoshenko** [comment]
I'll add something like it to `meta.description`.

**Comment by dtzWill** [comment]
According to [docs](https://developer.gnome.org/glib/stable/glib-Message-Logging.html#g-error) it seems `g_error` calls `abort` instead of `exit` which results in a coredump and is not intended for user-facing errors.

This seems inappropriate for at least the error path that prints messages to the user about running `sysctl`.

Maybe make a pass with this in mind, or instead adjust the g_perr macro to just log and exit instead?

**Comment by yegortimoshenko** [comment]
`SIGTRAP` is correct. Also, failed `unshare` might be a user configuration problem, or might be not, and coredump might be useful.

**Comment by dtzWill** [comment]
Not to bikeshed but I can't help but wonder how much of benefit glib brings here compared to C++ or something (g_autofree patterns, etc.)?

Regardless I'd appreciate, if for my curiosity if nothing else, a mention of the role glib plays in obtaining the improvements in this PR or to what extent they're orthogonal?

**Comment by yegortimoshenko** [comment]
I think C++ is a bad language because of very complex semantics.

`g_autofree` is negligible convenience, and improvement mostly comes from utility functions: `g_build_filename`, `g_mkdtemp_full` (handles `chmod`), `g_file_test`, `g_dir_open` (doesn't return `.` and `..` unlike `opendir`), `g_strv_contains`.

**Comment by dtzWill** [comment]
Okay, thanks for the comments and consideration of abort.  LGTM review-wise.

**Comment by yegortimoshenko** [comment]
Thanks for reviewing it!

File: pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+#include <glib.h>
+#include <glib/gstdio.h>
+
+#include <errno.h>
+#include <sched.h>
+#include <unistd.h>
+
+#define g_perr(s, err) g_error("%s#L%d: %s: %s", __func__, __LINE__, s, g_strerror(err))
+#define g_perrno(s) g_perr(s, errno)
```
Comment (orivej) [inline pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: `g_` is a glib namespace, you should not define names starting with `g_`.

Comment (yegortimoshenko) [reply pkgs/build-support/build-fhs-userenv/chrootenv/chrootenv.c]: Sure, and names themselves are also bad and not descriptive. I've pushed a new version with `fail` and `fail_if` macros: I think it separates concerns much better than previously used error macros, and names are much more straightforward.

**Review by Mic92 — APPROVED** [review]
I have not tested the code itself, but the rest looks good to me.

**Comment by yegortimoshenko** [comment]
@Mic92: Thanks for mentioning `__cleanup__` GCC/Clang attribute!

**Comment by yegortimoshenko** [comment]
@orivej: Thanks for the review!

I still want to implement a different order of messages in case of failed `unshare`, and the only clean way that I could come up with is via a SIGABRT or SIGTRAP handler. The problem is that if I use GLib logging from the handler, call to any logging function results in recursion warning and a subsequent abort.

**Pull request merged**